File: [I]stack-queue/[I]stack-queue/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        close_open = {')': '(', '}': '{', ']': '['}
        stack = []
        for c in s:
            if c not in close_open:
                stack.append(c)
            else:
                if not stack or stack[- 1] != close_open[c]:
                    return False
                stack.pop()
        return not stack

# time O(n)
# space O(n)
# using stack and queue and use stack to store the last states and hashmap

File: [I]stack-queue/[I]stack-queue/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        HashMap<Character, Character> closeOpen = new HashMap<>();
        closeOpen.put(')', '(');
        closeOpen.put('}', '{');
        closeOpen.put(']', '[');
        ArrayDeque<Character> stack = new ArrayDeque<>();
        for (Character c: s.toCharArray()) {
            if (closeOpen.containsKey(c)) {
                if (!closeOpen.get(c).equals(stack.peek())) {
                    return false;
                }
                stack.pop();
            } else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}

// time O(n)
// space O(n)
// using stack and queue and use stack to store the last states and hashmap